Add initial implementation (#1)
vshn/statefulset-resize-controller: ARCHITECTURE.md

# Architecture

This document describes the high-level architecture of the Statefulset Resize Controller.
It should give you a good impression on *how* it works and *where* to change things.

## Overview

On a very high-level, the Statefulset Resize Controller watches changes on Statefulsets and reacts to them.
This is the general way Kubernetes Operators work and we call this reacting to resource changes the *reconcile loop*.
If you are not familiar with Kubernetes Operators it might be helpful to read up on them before proceeding.

> The controller will notice Statefulsets that should be resized, make a copy of the too small volumes, recreate the original volumes with the requested size, and restore the original data from the backup.
> It does this by only interacting with the Kubernetes API and has no access to other systems.

Reviewer: ```suggestion
The controller will notice Statefulsets that should be resized, make a copy of the current volumes by creating new volumes with the requested size, delete the previous volume and then restore the original data from the previous volumes to the new volumes.
It does this by only interacting with the Kubernetes API and has no access to other systems.
```

│   ├── controller.go       # Entry point of reconcile loop
│   ├── statefulset.go      # Fetch and update Sts, initiate resize
│   ├── pvc.go              # Fetch relevant PVCs
│   ├── backup.go           # Create backup PVC and job
│   ├── restore.go          # Recreate PVC and restore data
│   └── copy.go             # Handle job creation
├── statefulset           # Wrapper handling modification of k8s Statefulset resources
├── pvc                   # Wrapper handling modification of k8s PVC resources
├── cmd                   # Entry point - main.go
└── naming                # General helper package for naming
```

vshn.net

## Resize Steps

There are five distinct steps that are part of a successful resize. 
Each of these steps is abstracted as an idempotent function that will return whether it completed or has completed successfully before.

All information on the state of the resize is stored as annotations of the statefulset.

### Find Resizable PVCs

TODO DIAGRAM 

The controller is notified of any statefulset change.
This means as the very first step we detect whether the statefulset needs to be resized.

> This first includes fetching the statefulset, and checking whether we are already in the process of resizing.
> It then finds all PVCs that are smaller then the PVC template of the statefulset.
> If there are any, they are stored on the statefulset and we proceed with the resizing.

Reviewer: I would keep the spelling the same as Kubernetes does (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) (suggestion applies to the rest of the document as well)
```suggestion
This first includes fetching the StatefulSet, and checking whether we are already in the process of resizing.
It then finds all PVCs that are smaller then the PVC template of the StatefulSet.
If there are any, they are stored on the StatefulSet and we proceed with the resizing.
```

> Fetching the statefulset is handled in `controllers/statefulset.go`. Finding PVCs is handled in `controller/pvc.go`.

Reviewer: ```suggestion
Fetching the statefulset is handled in `controllers/statefulset.go`. 
Finding PVCs is handled in `controller/pvc.go`.
```
